### PR TITLE
Extend mktemp to support file suffixes.

### PR DESCRIPTION
**Pull request by maleadt**
This PR extends `mktemp` to support file suffixes using the `mkstemps` function:

```
mktemp(suffix=".txt")
("/tmp/tmprFezJ1.txt",IOStream(<fd 15>))
```

Additionally, I've also changed the ccall signature from `Ptr{UInt8}` to `Cstring`. My understanding is that `Cstring` is meant to be used when a string is NULL-terminated, and according to [this document](https://www.mkssoftware.com/docs/man3/mkstemp.3.asp) the template parameter to `mkstemp` and others is supposed to be terminated as such.
Am I reading this correctly? If so, I think other functions (`tempnam`, `mktempdir`) might need adjusting too.

### CODE REVIEW

**Comment by tkelman** [comment]
needs tests, docs, and if possible a windows implementation (if not, an error)

**Comment by maleadt** [comment]
Updated. I force-pushed, because I reverted the change from `Ptr{UInt8}` to `Cstring`, as I'm going to keep that for another PR.

**Comment by vtjnash** [comment]
Can you explain the purpose of this PR? The Win32 API doesn't support suffixes (https://msdn.microsoft.com/en-us/library/windows/desktop/aa364991%28v=vs.85%29.aspx) so that would need to be reimplemented from scratch, and suffixes are not used much on unix systems so they should not have much value there.

**Comment by maleadt** [comment]
I was using Julia to interface with an application whose behaviour depends on the file extension, hence I needed to create temporary files with a specific suffix. Didn't seem unreasonable to me, and I had used Python's `tempfile.TemporaryFile` before which does support such suffixes (cpython completely reimplements this logic, not relying on OS APIs).

I botched the Windows version though, not understanding `tempname` _creates_ the file on Windows. One option would be to rename the file, replacing the `.tmp` it is guaranteed to contain to the requested suffix. But then the file isn't guaranteed to be unique; just add a check, or really reimplement from scratch?

Or I just close this PR and don't add support for suffixes...

**Comment by tkelman** [comment]
ref #9053, still need to fix that

File: test/file.jl [hunk 1]
```diff
@@ -219,6 +219,12 @@ close(f)
 @test readall(p) == "Here is some text"
 rm(p)
 
+(p, f) = mktemp(suffix=".foo")
+close(f)
+@test isfile(p) == true
+@test endswith(p, ".foo") == true
+rm(p)
```
Comment (hayd) [inline test/file.jl]: You can use do block in mktemp to avoid both close and rm.

File: base/file.jl [hunk 2]
```diff
@@ -188,9 +188,13 @@ end
 tempdir() = dirname(tempname())
 
 # Create and return the name of a temporary file along with an IOStream
-function mktemp(parent=tempdir())
-    b = joinpath(parent, "tmpXXXXXX")
-    p = ccall(:mkstemp, Int32, (Ptr{UInt8},), b) # modifies b
+function mktemp(parent=tempdir(); suffix="")
+    b = joinpath(parent, "tmpXXXXXX$suffix")
+    if length(suffix) > 0
+        p = ccall(:mkstemps, Int32, (Ptr{UInt8}, Cint), b, length(suffix))
```
Comment (hayd) [inline base/file.jl]: Why not just always use mktemps ?

Comment (maleadt) [inline base/file.jl]: Makes sense, I'll update it.

**Pull request closed without merging**